core/iwasm/libraries/lib-socket/src/wasi/wasi_socket_ext.c

```diff
@@ -282,3 +282,80 @@ getpeername(int sockfd, struct sockaddr *addr, socklen_t *addrlen)
 
     return __WASI_ERRNO_SUCCESS;
 }
+
+struct timeval
+time_us_to_timeval(uint64_t time_us)
+{
+    struct timeval tv;
+    tv.tv_sec = time_us / 1000000UL;
+    tv.tv_usec = time_us % 1000000UL;
+    return tv;
+}
+
+uint64_t
+timeval_to_time_us(struct timeval tv)
+{
+    return (tv.tv_sec * 1000000UL) + tv.tv_usec;
+}
+
+int
+getsockopt(int sockfd, int level, int optname, void *__restrict optval,
+           socklen_t *__restrict optlen)
+{
+    __wasi_errno_t error;
+    uint64_t timeout_us;
+
+    switch (level) {
+        case SOL_SOCKET:
+            switch (optname) {
```

Comment:
`switch` inside another `switch` makes the code a bit less readable, consider refactoring the code to avoid deeply nested code.

Reply:
Agree, maybe separate functions depending on the level? What'd you recommend?

Reply:
I think splitting it into functions is not a bad idea. Also, I don't expect we're going to add support to level different than `SOL_SOCKET`, so you might consider failing early at the beginning of the function:
```
if (level != SOL_SOCKET) {
  return -1;
}
```

Reply:
Thanks, will split because I was planning to include a couple `IPPROTO_TCP` ones